pkg/utils/object.go

```diff
@@ -0,0 +1,224 @@
+// The object is created or updated with the given content, except for system metadata fields, namespace, and name.
+// This function can be combined with runtime.DefaultUnstructuredConverter.ToUnstructured to create or update an object
+// from runtime.RawExtension.
+func CreateOrUpdateObject(ctx context.Context, c client.Client, gvk schema.GroupVersionKind, name, namespace string, content map[string]interface{}) error {
+	// Initialize the object
+	key := client.ObjectKey{Namespace: namespace, Name: name}
+	obj := &unstructured.Unstructured{}
+	obj.SetGroupVersionKind(gvk)
+	obj.SetName(name)
+	obj.SetNamespace(namespace)
+
+	// Create or update the object with retries for optimistic concurrency
+	for retries, done := 0, false; !done; retries++ {
+		// Check if the object already exists
+		found := true
+		if err := c.Get(ctx, key, obj); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return errors.Wrapf(err, "could not get object %s %s", gvk, key)
+			}
+
+			// The object was not found
+			found = false
+		}
+
+		// Set object content
+		if content != nil {
+			obj.SetUnstructuredContent(mergeObjectContents(obj.UnstructuredContent(),
+				filterMetadata(content, add(systemMetadataFields, "namespace", "name")...)))
+		}
+
+		// Create or update the object
+		var err error
+		if !found {
+			err = c.Create(ctx, obj)
+		} else {
+			err = c.Update(ctx, obj)
+		}
+		if err != nil && (!apierrors.IsConflict(err) || retries == maxRetries) {
+			return errors.Wrapf(err, "could not create or update object %s %s", gvk, key)
+		}
+
+		done = err == nil
+	}
+	return nil
+}
+
+// DeleteObjectByRef deletes the object with the given reference and namespace using the given client.
+func DeleteObjectByRef(ctx context.Context, c client.Client, ref *autoscalingv1.CrossVersionObjectReference, namespace string) error {
+	gvk, err := gvkFromCrossVersionObjectReference(ref)
+	if err != nil {
+		return err
+	}
+	return DeleteObject(ctx, c, gvk, ref.Name, namespace)
+}
+
+// DeleteObject deletes the object with the given GVK, name, and namespace using the given client.
+func DeleteObject(ctx context.Context, c client.Client, gvk schema.GroupVersionKind, name, namespace string) error {
+	// Initialize the object
+	key := client.ObjectKey{Namespace: namespace, Name: name}
+	obj := &unstructured.Unstructured{}
+	obj.SetGroupVersionKind(gvk)
+
+	// Delete the object with retries for optimistic concurrency
+	for retries, done := 0, false; !done; retries++ {
+		// Get the object
+		if err := c.Get(ctx, key, obj); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return errors.Wrapf(err, "could not get object %s %s", gvk, key)
+			}
+
+			// The object was not found
+			return nil
+		}
+
+		// Delete the object
+		err := c.Delete(ctx, obj)
+		if err != nil && (!apierrors.IsConflict(err) || retries == maxRetries) {
+			return errors.Wrapf(err, "could not delete object %s %s", gvk, key)
+		}
+
+		done = err == nil
+	}
+	return nil
```

Comment:
Wouldn't

```suggestion
	return client.IgnoreNotFound(c.Delete(ctx, obj))
```

be enough? How can a conflict happen during delete?

Reply:
It seems a conflict can't happen during delete, sorry, my bad. Done as suggested.